### DISCUSSION TITLE
Do NOT share history

### DISCUSSION THREAD

**Discussion by giacgbj** [ohmyzsh/ohmyzsh#9820] [Help] [answered]
I've already read and tested all the possible combinations of parameters suggested in many tickets found here (e.g.: #2537) but nothing prevents the history to be shared among windows/tabs after closing/reopening iTerm2 on macOS Catalina.

The only way I found (elsewhere) is:
`HISTFILE=~/.zsh-history-${ITERM_SESSION_ID}`

Is It the right way? Could this have any drawback?

**Comment by mcornella** [comment, marked as answer]
It should be enough to add `unsetopt share_history` to the zshrc file, after OMZ is sourced.

EDIT: behavior doesn't align with user description, see https://github.com/ohmyzsh/ohmyzsh/discussions/9820#discussioncomment-604161.

**Reply by giacgbj** [reply]
It doesn't work. I even tried commenting out `setopt share_history` in `~/.oh-my-zsh/lib/history.zsh`

**Reply by mcornella** [reply]
Can you post the zshrc file?  Also, the behavior should be as follows:

- tab 1 and tab 2 are running side by side
- history in tab 1 doesn't appear in tab 2
- when you close a tab, the history of that tab is written and available to new tabs

Is that not happening?

**Reply by rwmitchell** [reply]
Their problem is not with the zsh share history feature, but that when exiting a shell, zsh (as most shells do) save their history into a single file.  When a new window is opened, it loads the same single file.  Their solution is to use the iterm2 id to create multiple history files, one for each unique iterm2 id.  This will work, or appear to, fairly well.  The issue will come when closing/opening windows and effectively getting a semi-random history loaded back in.

The iterm2 id tracks desktop, window, pane, so it should mostly work.

**Reply by mcornella** [reply]
So the history of a pane will never be used in any other tab, if I understand correctly? That's like saying that the history will be deleted once you close the pane, except it will be recorded in a file for later inspection with less, grep, etc.

**Reply by rwmitchell** [reply]
A full session id looks like:  ```w8t4p1:710814BD-8703-47E0-9BA4-C3A59F90E8E3```
Assuming only the first portion (before the colon) is used, histories are likely to be reused, but not necessarily in a way that is helpful.  A lot depends on how one works.

**Reply by giacgbj** [reply]
If I understand correctly, the behavior I'd like is the one described by @rwmitchell: I want separate histories for each tab. All the session id is used to create the history files (e.g.: `.zsh-history-w0t0p0:4624F72F-8D29-4F8B-9D37-26BACBD17B04`) and it seems to work.

**Reply by mcornella** [reply]
Great, then @rwmitchell you can post this as a separate answer so we can mark this as solved. GitHub doesn't yet allow to mark replies as answers.

**Reply by rwmitchell** [reply]
I could be wrong, haven't looked closely, but I'd highly suspect the portion after the : is unique.

I use ```sid=`echo $ITERM_SESSION_ID | sed 's/:.*//' ``` in another script to track which font used in each terminal.  I would use 'sid' to make the history files unique to just the first part.
